packages/opentelemetry-core/src/utils/deep-merge.ts

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+export function deepMerge(
+  target: Record<string, any>,
+  source: Record<string, any>
+) {
+  const merged = target;
+  for (const prop in source) {
+    if (typeof source[prop] === 'object' && source[prop] !== null) {
+      merged[prop] = deepMerge(target[prop], source[prop]);
+    } else {
+      merged[prop] = source[prop];
+    }
+  }
+  return merged;
+}
```

Comment:
This function does no sort of cycle checking, which could cause an infinite CPU loop. This needs at the very least to be documented, but it might be nice to actually do some cycle checking. You don't need to do anything fancy like a Floyd's algorithm, but a simple `maxDepth` optional parameter that defaults to 10 or something and decrements each time you recurse might be a good sanity check.

Reply:
Thanks, TIL

maxDepth param added... Can you provide a scenario where this could happen btw?